.github/workflows/typecheck.yml

```diff
@@ -0,0 +1,53 @@
+# Static type check
+#
+# This workflow runs static type check.
+#
+# It is run on every commit to the main and pull request branches. It is also
+# scheduled to run daily on the main branch.
+#
+name: Static Type Check
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+  # Schedule daily tests
+  schedule:
+    - cron: '0 0 * * *'
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: ${{ github.ref != 'refs/heads/main' }}
+
+jobs:
+  static_check:
+    name: Static Type Check
+    runs-on: ubuntu-latest
+
+    steps:
+      # Checkout current git repository
+      - name: Checkout
+        uses: actions/checkout@v4.1.1
+
+      # Setup Python
+      - name: Set up Python
+        uses: actions/setup-python@v4.7.1
+        with:
+          python-version: '3.12'
+
+      - name: Install packages
+        run: |
+          # Need to install four groups of packages:
+          # 1. required packages
+          # 2. optional packages
+          # 3. type checker and stub package
+          # 4. other packages that are used somewhere in PyGMT
+          python -m pip install \
+            numpy pandas xarray netcdf4 packaging \
+            contextily geopandas ipython rioxarray \
+            mypy pandas-stubs \
+            matplotlib pytest
+          python -m pip list
```

Comment:
Haven't used this before, but could we use [`mypy --install-types ...`](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-install-types) following https://github.com/pydata/xarray/blob/f7edd1e8c2d6d1a8668cc8baba9e1c66d87b8bb4/.github/workflows/ci-additional.yaml#L124C14-L124C14 instead of listing all these dependencies?

Reply:
It doesn't work. The doc says:

> This flag causes mypy to install known missing stub packages for third-party libraries using pip.

So it only works for pandas which maintains a separate stub package `pandas-stubs` and can't install packages like numpy.

I tried to create a fresh environment and run 
```
$ conda create --name test python=3.12 mypy
$ conda activate test
$ mypy pygmt
pygmt/io.py:4: error: Cannot find implementation or library stub for module named "xarray"  [import-not-found]
pygmt/datasets/tile_map.py:7: error: Cannot find implementation or library stub for module named "contextily"  [import-not-found]
pygmt/datasets/tile_map.py:11: error: Cannot find implementation or library stub for module named "numpy"  [import-not-found]
...
# A lot of similar errors 

Installing missing stub packages: python -m pip install pandas-stubs

Install? [yN]
```